ably/proto_message.go

```diff
@@ -193,6 +241,116 @@ func (m Message) withDecodedData(cipher channelCipher) (Message, error) {
+	// strings.Split on empty string returns []string{""}
+	if m.Data == nil || m.Encoding == "" {
+		return m, nil, nil
+	}
+
+	var basePayloadForReturn []byte
+	encodings := strings.Split(m.Encoding, "/")
+
+	for len(encodings) > 0 {
+		encoding := encodings[len(encodings)-1]
+		encodings = encodings[:len(encodings)-1]
+		switch encoding {
+		case encVCDiff:
+			// Handle vcdiff delta decoding (PC3, RTL18, RTL19, RTL20)
+			if ctx == nil || ctx.VCDiffPlugin == nil {
+				return m, nil, fmt.Errorf("missing VCdiff decoder plugin (code 40019)")
+			}
+
+			// Validate delta reference ID (RTL20)
+			deltaExtras := extractDeltaExtras(m.Extras)
+			if deltaExtras != nil && deltaExtras.From != ctx.LastMessageID {
+				return m, nil, fmt.Errorf("delta message decode failure - delta reference ID %q does not match stored ID %q (code 40018)", deltaExtras.From, ctx.LastMessageID)
+			}
+
+			if ctx.BasePayload == nil {
+				return m, nil, fmt.Errorf("delta message decode failure - no base payload available (code 40018)")
+			}
+
+			deltaBytes, err := coerceBytes(m.Data)
+			if err != nil {
+				return m, nil, err
+			}
+
+			// Decode the delta
+			result, err := ctx.VCDiffPlugin.Decode(deltaBytes, ctx.BasePayload)
+			if err != nil {
+				return m, nil, fmt.Errorf("vcdiff decode failed: %w (code 40018)", err)
+			}
+
+			m.Data = result
+			basePayloadForReturn = result // Store as new base payload (RTL19c)
+
+		case encBase64:
+			d, err := coerceString(m.Data)
+			if err != nil {
+				return m, nil, err
+			}
+			data, err := base64.StdEncoding.DecodeString(d)
+			if err != nil {
+				return m, nil, err
+			}
+			m.Data = data
+			// For non-delta messages, store as base payload after base64 decoding (RTL19a)
+			if basePayloadForReturn == nil {
+				basePayloadForReturn = data
+			}
+
+		case encUTF8:
+			d, err := coerceString(m.Data)
+			if err != nil {
+				return m, nil, err
+			}
+			m.Data = d
+
+		case encJSON:
+			d, err := coerceBytes(m.Data)
+			if err != nil {
+				return m, nil, err
+			}
+			var result interface{}
+			if err := json.Unmarshal(d, &result); err != nil {
+				return m, nil, fmt.Errorf("error unmarshaling JSON payload of type %T: %s", m.Data, err.Error())
+			}
+			m.Data = result
+
+		default:
+			if strings.HasPrefix(encoding, encCipher) {
+				if cipher == nil {
+					return m, nil, fmt.Errorf("message data is encrypted as %s, but cipher wasn't provided", encoding)
+				}
+				d, err := coerceBytes(m.Data)
+				if err != nil {
+					return m, nil, err
+				}
+				d, err = cipher.Decrypt(d)
+				if err != nil {
+					return m, nil, fmt.Errorf("decrypting message data: %w", err)
+				}
+				m.Data = d
+			} else {
+				return m, nil, fmt.Errorf("unknown encoding %s", encoding)
+			}
+		}
+		m.Encoding = strings.Join(encodings, "/")
+	}
+
+	// For non-delta messages, store the final decoded data as base payload (RTL19b)
+	if basePayloadForReturn == nil {
```

Comment:
It's a bit unclear what this is for, doesn't the returned Message.Data field have the decoded data?